File: ex214.java
class Solution {
    public String shortestPalindrome(String s) {        
        int len = s.length();
        
        int end = len;
        int i = 0;
        int j = end - 1;
        
        while (i < j) {
            if (s.charAt(i) == s.charAt(j)) {
                i++;    j--;
            } else {
                i = 0;    end--;    j = end - 1;
            }
        }
        
        return new StringBuilder(s.substring(end, len)).reverse().toString() + s;
    }
}

File: code/ex214.cpp
class Solution {
public:
    string shortestPalindrome(string s) {
        int n = s.length();
        
        string rev = s;
        reverse(rev.begin(), rev.end());
        string str = s + " " + rev;
        
        vector<int> nxt = getNext(str);
        
        return rev.substr(0, rev.length() - nxt.back()) + s;
    }
    
private:
    vector<int> getNext(string s) {
        int n = s.length();
        vector<int> nxt{0, 0};
        for (int i = 1, j = 0; i < n; i++) {
            while (j > 0 && s[i] != s[j])    j = nxt[j];
            if (s[i] == s[j])    j++;
            nxt.push_back(j);
        }
        return nxt;
    }
};